presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,169 @@
+import com.google.common.collect.ListMultimap;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.getExpressionTypes;
+import static com.facebook.presto.type.UnknownType.UNKNOWN;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Ensures that all the expressions and FunctionCall's matches their output symbols
+ */
+public final class TypeMatchingChecker
+        implements PlanSanityChecker.Checker
+{
+    public TypeMatchingChecker() {}
+
+    @Override
+    public void validate(Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types, PlanNode plan)
+    {
+        plan.accept(new Visitor(session, metadata, sqlParser, types), null);
+    }
+
+    private static class Visitor
+            extends SimplePlanVisitor<Void>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final SqlParser sqlParser;
+        private final Map<Symbol, Type> types;
+
+        public Visitor(Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.types = requireNonNull(types, "types is null");
+        }
+
+        @Override
+        public Void visitAggregation(AggregationNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            AggregationNode.Step step = node.getStep();
+
+            // Do checking only when the step is SINGLE since FINAL and PARTIAL aggregation might have intermediate representation types
+            if (step.equals(AggregationNode.Step.SINGLE)) {
+                checkSignatureTypeMatching(node.getFunctions());
+                checkFunctionCallTypeMatching(node.getAggregations());
+            }
+
+            return null;
+        }
+
+        @Override
+        public Void visitWindow(WindowNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkSignatureTypeMatching(node.getSignatures());
+            checkFunctionCallTypeMatching(node.getWindowFunctions());
+
+            return null;
+        }
+
+        @Override
+        public Void visitProject(ProjectNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkExpressionTypeMatching(node.getAssignments());
+
+            return null;
+        }
+
+        @Override
+        public Void visitUnion(UnionNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkMappingTypeMatching(node.getSymbolMapping());
+
+            return null;
+        }
+
+        private void checkMappingTypeMatching(ListMultimap<Symbol, Symbol> symbolMapping)
```

Comment:
This seems to be only have one callsite. Maybe don't factor it out into a separate method?